core/drivers/clk/clk.c

```diff
@@ -0,0 +1,299 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Bootlin
+ */
+
+#include <drivers/clk.h>
+#include <kernel/boot.h>
+#include <kernel/panic.h>
+#include <kernel/spinlock.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <stddef.h>
+
+/* Global clock tree lock */
+static unsigned int clk_lock = SPINLOCK_UNLOCK;
+
+struct clk *clk_alloc(const char *name, const struct clk_ops *ops,
+		      struct clk **parent_clks, size_t parent_count)
+{
+	struct clk *clk = NULL;
+	size_t parent = 0;
+
+	clk = calloc(1, sizeof(*clk) + parent_count * sizeof(clk));
+	if (!clk)
+		return NULL;
+
+	if (parent_count) {
+		clk->num_parents = parent_count;
+		for (parent = 0; parent < parent_count; parent++)
+			clk->parents[parent] = parent_clks[parent];
+	}
+
+	clk->name = name;
+	clk->ops = ops;
+	refcount_set(&clk->enabled_count, 0);
+
+	return clk;
+}
+
+void clk_free(struct clk *clk)
+{
+	free(clk);
+}
+
+static bool clk_check(struct clk *clk)
+{
+	if (!clk->ops)
+		return false;
+
+	if (clk->ops->set_parent && !clk->ops->get_parent)
+		return false;
+
+	if (clk->num_parents > 1 && !clk->ops->get_parent)
+		return false;
+
+	return true;
+}
+
+static void clk_compute_rate_no_lock(struct clk *clk)
+{
+	unsigned long parent_rate = 0;
+
+	if (clk->parent)
+		parent_rate = clk->parent->rate;
+
+	if (clk->ops->get_rate)
+		clk->rate = clk->ops->get_rate(clk, parent_rate);
+	else
+		clk->rate = parent_rate;
+}
+
+struct clk *clk_get_parent_by_index(struct clk *clk, size_t pidx)
+{
+	if (pidx >= clk->num_parents)
+		return NULL;
+
+	return clk->parents[pidx];
+}
+
+static void clk_init_parent(struct clk *clk)
+{
+	size_t pidx = 0;
+
+	switch (clk->num_parents) {
+	case 0:
+		break;
+	case 1:
+		clk->parent = clk->parents[0];
+		break;
+	default:
+		pidx = clk->ops->get_parent(clk);
+		assert(pidx < clk->num_parents);
+
+		clk->parent = clk->parents[pidx];
+		break;
+	}
+}
+
+TEE_Result clk_register(struct clk *clk)
+{
+	assert(clk_check(clk));
+
+	/*
+	 * No need to lock from here, the child is not yet added to the clock
+	 * tree.
+	 */
+	clk_init_parent(clk);
+
+	cpu_spin_lock(&clk_lock);
+	clk_compute_rate_no_lock(clk);
+	cpu_spin_unlock(&clk_lock);
+
+	DMSG("Registered clock %s, freq %ld", clk->name, clk_get_rate(clk));
+
+	return TEE_SUCCESS;
+}
+
+static bool clk_is_enabled_no_lock(struct clk *clk)
+{
+	return refcount_val(&clk->enabled_count) != 0;
+}
+
+static struct clk *clk_get_parent_no_lock(struct clk *clk)
+{
```

Comment:
if this function is not planned to evolve, maybe better use ` = clk->parent` where called.

Reply:
Ok, I removed it since the lock was also removed.